File: app/src/main/java/com/pnuema/bible/ui/fragments/VerseSelectionFragment.java
package com.pnuema.bible.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.pnuema.bible.R;
import com.pnuema.bible.data.firefly.VerseCount;
import com.pnuema.bible.retrievers.BaseRetriever;
import com.pnuema.bible.retrievers.FireflyRetriever;
import com.pnuema.bible.statics.CurrentSelected;
import com.pnuema.bible.ui.adapters.NumberSelectionAdapter;
import com.pnuema.bible.ui.dialogs.BCVSelectionListener;
import com.pnuema.bible.ui.dialogs.NumberSelectionListener;

import java.util.Observable;
import java.util.Observer;

/**
 * A fragment representing a list of verse numbers to pick from.
 * <p/>
 */
public class VerseSelectionFragment extends Fragment implements Observer, NumberSelectionListener {
    private BCVSelectionListener mListener;
    private BaseRetriever mRetriever = new FireflyRetriever();
    private RecyclerView mGridView;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public VerseSelectionFragment() {
    }

    @Override
    public void setUserVisibleHint(final boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        mRetriever.addObserver(this);
        if (isVisibleToUser && CurrentSelected.getChapter() != null) {
            mRetriever.getVerseCount(CurrentSelected.getVersion(), String.valueOf(CurrentSelected.getBook()), String.valueOf(CurrentSelected.getChapter()));
        }
    }

    @SuppressWarnings("unused")
    public static VerseSelectionFragment newInstance(final BCVSelectionListener listener) {
        return new VerseSelectionFragment().setListener(listener);
    }

    private VerseSelectionFragment setListener(final BCVSelectionListener listener) {
        mListener = listener;
        return this;
    }

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        mGridView = (RecyclerView) inflater.inflate(R.layout.fragment_number_list, container, false);
        mGridView.setLayoutManager(new GridLayoutManager(getContext(), 3));
        return mGridView;
    }

    @Override
    public void onPause() {
        super.onPause();
        mRetriever.deleteObservers();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void update(final Observable observable, final Object o) {
        final Context context = getContext();
        if (context == null || getActivity() == null || getActivity().isFinishing()) {
            return;
        }

        if (o instanceof VerseCount) {
            mGridView.setAdapter(new NumberSelectionAdapter(((VerseCount)o).getVerseCount(), CurrentSelected.getVerse(), this));
        }
    }

    @Override
    public void numberSelected(final int number) {
        mListener.onVerseSelected(number);
    }
}

File: app/src/main/java/com/pnuema/bible/android/ui/fragments/VerseSelectionFragment.kt
package com.pnuema.bible.android.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pnuema.bible.android.R
import com.pnuema.bible.android.data.firefly.VerseCount
import com.pnuema.bible.android.statics.CurrentSelected
import com.pnuema.bible.android.ui.adapters.NumberSelectionAdapter
import com.pnuema.bible.android.ui.dialogs.BCVSelectionListener
import com.pnuema.bible.android.ui.dialogs.NumberSelectionListener
import com.pnuema.bible.android.ui.fragments.viewModel.VersesViewModel

/**
 * A fragment representing a list of verse numbers to pick from.
 *
 *
 */
class VerseSelectionFragment(private val listener: BCVSelectionListener) : Fragment(R.layout.fragment_number_list), NumberSelectionListener {
    private lateinit var viewModel: VersesViewModel

    companion object {
        fun newInstance(listener: BCVSelectionListener): VerseSelectionFragment {
            return VerseSelectionFragment(listener)
        }
    }

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    init {
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val gridView = view as RecyclerView
        gridView.layoutManager = GridLayoutManager(context, 3)

        viewModel = ViewModelProvider(this).get(VersesViewModel::class.java)
        viewModel.verses.observe(viewLifecycleOwner, Observer { verseCount ->
            if (activity == null || activity?.isFinishing != false) {
                return@Observer
            }

            if (verseCount is VerseCount) {
                gridView.adapter = NumberSelectionAdapter(verseCount.verseCount, CurrentSelected.verse, this)
            }
        })
    }

    override fun onResume() {
        super.onResume()

        if (isVisible && CurrentSelected.chapter != CurrentSelected.DEFAULT_VALUE) {
            viewModel.loadVerses(CurrentSelected.version, CurrentSelected.book.toString(), CurrentSelected.chapter.toString())
        }
    }

    override fun numberSelected(number: Int) {
        listener.onVerseSelected(number)
    }
}
